datastream-server/src/main/java/com/linkedin/datastream/server/EventProducer.java

```diff
@@ -201,26 +201,28 @@ private void validateEventRecord(DatastreamProducerRecord record) {
   }
 
   @Override
-  public void broadcast(DatastreamProducerRecord record, SendCallback sendBroadcastCallback) {
-    helperSendOrBroadcast(record, sendBroadcastCallback, null, true);
+  public DatastreamRecordMetadata broadcast(DatastreamProducerRecord record, SendCallback sendEventCallback) {
+    return helperSendOrBroadcast(record, sendEventCallback, true);
   }
 
   @Override
   public void send(DatastreamProducerRecord record, SendCallback sendCallback) {
-    helperSendOrBroadcast(record, null, sendCallback, false);
+    helperSendOrBroadcast(record, sendCallback, false);
   }
 
   /**
    * Send the event onto the underlying transport.
    * @param record the datastream event
-   * @param sendBroadcastCallback the callback to be invoked after the event is broadcasted to the destination
    * @param sendEventCallback the callback to be invoked after the event is sent to the destination
+   *
+   * @return For broadcast return DatastreamRecordMetadata got from transport provider broadcast, null for send
    */
-  public void helperSendOrBroadcast(DatastreamProducerRecord record, SendCallback sendBroadcastCallback,
+  private DatastreamRecordMetadata helperSendOrBroadcast(DatastreamProducerRecord record,
       SendCallback sendEventCallback, boolean isBroadcast) {
+    DatastreamRecordMetadata broadcastMetadata = null;
+
     try {
       validateEventRecord(record);
-
       try {
```

Comment:
nested try? can probably simplify and have one try-catch pair per method.

Reply:
It seems a little problematic to refactor this in a neat way.

The try block for serialization error has a catch all Exception and I wan't able to identify the correct set of specific exceptions that might occur, so I introduced a flag to differentiate the exception in the catch block. Other alternative if to have multiple try blocks instead of nested try block. I am personally not partial to either adding a flag or nested try block. But pls let me know if there is a cleaner way that I might have overlooked.